include/wlr/types/wlr_output.h

```diff
@@ -303,26 +294,90 @@ void wlr_output_lock_attach_render(struct wlr_output *output, bool lock);
+ * draw, you should fall back to using software cursors.
+ *
+ * This returns false if the backend does not support hardware cursors or from
+ * some other internal error. Either way, you should fall back to using
+ * software cursors.
+ */
+bool wlr_output_cursor_try_set_size(struct wlr_output *output, int *x, int *y);
+
+/**
+ * Attaches the renderer to the cursor buffer.
+ *
+ * You must successfully call `wlr_output_cursor_try_set_size` at least once
+ * before using this. Any rendering done will be applied on
+ * `wlr_output_cursor_commit`.
+ */
+bool wlr_output_cursor_attach_render(struct wlr_output *output, int *buffer_age);
 
-struct wlr_output_cursor *wlr_output_cursor_create(struct wlr_output *output);
 /**
- * Sets the cursor image. The image must be already scaled for the output.
+ * Enables displaying the hardware cursor.
+ *
+ * This does not affect the ability to render to the cursor. It only sets
+ * whether or not it will be displayed. By default, the cursor is disabled.
+ *
+ * This will be applied on `wlr_output_cursor_commit`.
+ */
+void wlr_output_cursor_enable(struct wlr_output *output, bool enable);
+
+/**
+ * Moves the position of the hardware cursor on the main framebuffer.
+ *
+ * This is in output buffer coordinates. Negative values and values that would
+ * place the cursor partially or entirely off the main framebuffer are allowed.
+ *
+ * The hotspot is the point on the cursor buffer (from the top left) where the
+ * cursor actually points. For example, with a cursor representing a hand, the
+ * hotspot may be over the index finger, rather than the top left corner.
+ *
+ * Some backends may not be able to move the cursor freely, but may still be
+ * able to make use of the hotspot to correctly show the cursor.
+ *
+ * This will be applied on `wlr_output_cursor_commit`.
+ */
+void wlr_output_cursor_move(struct wlr_output *output, int x, int y,
+	int hotspot_x, int hotspot_y);
+
+/**
+ * Apply pending cursor state.
+ *
+ * This may or may not be synchronized with the output, and would be applied on
+ * `wlr_output_commit`, depending on the backend limitations. However, it's
+ * advised to assume that they are sychronized and do any cursor operations at
+ * the same time you render to the output.
+ *
+ * If you have called `wlr_output_cursor_attach_render`, you must call this
+ * function before `wlr_output_attach_render`, otherwise you risk losing or
+ * corrupting rendering state.
  */
-bool wlr_output_cursor_set_image(struct wlr_output_cursor *cursor,
-	const uint8_t *pixels, int32_t stride, uint32_t width, uint32_t height,
-	int32_t hotspot_x, int32_t hotspot_y);
-void wlr_output_cursor_set_surface(struct wlr_output_cursor *cursor,
-	struct wlr_surface *surface, int32_t hotspot_x, int32_t hotspot_y);
-bool wlr_output_cursor_move(struct wlr_output_cursor *cursor,
-	double x, double y);
-void wlr_output_cursor_destroy(struct wlr_output_cursor *cursor);
+bool wlr_output_cursor_commit(struct wlr_output *output);
```

Comment:
Any reason why we need an explicit function for this, instead of doing the commit with `wlr_output_commit`?

Reply:
Originally this is what I was I was going to do, but this caused issues with the rendering state.
```c
eglMakeCurrent(egl, cursor_surface, ...);
// render stuff
eglMakeCurrent(egl, primary_surface, ...);
// render stuff
eglSwapBuffers(egl, cursor_surface, ...);
eglSwapBuffers(egl, primary_surface, ...);
```
doesn't really work, even if you add additional eglMakeCurrent calls before eglSwapBuffers.

Reply:
I wonder if we should move to `begin/end` semantics, rather than `current/commit`.